book.md

```diff
@@ -510,6 +510,154 @@ can also implement automatic bounds checking for arrays where the array bound
+the sake of completeness, a very simple example code is shown below.
+
+::: {.example ex:use-after-free
+     caption="Simple UaF example"}
+
+```c
+struct auth_t {
+  char name[32];
+  int logged_in;
+};
+ 
+int main(int argc, char** argv) {
+  char line[50];
+ 
+  while(1) {
+    printf("[ auth = %p, service = %p ]\n", auth, service);
+ 
+    if (fgets(line, sizeof(line), stdin) == NULL) break;
+ 
+    // Usage: auth <name>
+    if (strncmp(line, "auth ", 5) == 0) {
+      auth = (struct auth_t*)malloc(sizeof(struct auth_t));
+
+      // Memory is only set to 0 here, not on free[1]
+      memset(auth, 0, sizeof(struct auth_t));
+
+      if (strlen(line + 5) < 31) {
+        strcpy(auth->name, line + 5);
+      }
+    }
+ 
+    // Usage: reset
+    else if (strncmp(line, "reset", 5) == 0) {
+      // [1]
+      free(auth);
+    }
+
+    // Usage: service <service-name>
+    else if (strncmp(line, "service ", 8) == 0) {
+      service = strdup(line + 8);
+    }
+
+    // Usage: login
+    else if (strncmp(line, "login", 5) == 0) {
+      // Possible use-after-free:
+      if (auth && auth->logged_in) {
+        printf("You are already logged in!\n");
+      } else {
+        printf("NOT AUTHORIZED!\n");
+      }
+    }
+  }
+
+  return 0;
+}
+```
+
+
+The exploit below may not work in every system, as it assumes that
+calling malloc+free+malloc will return the same pointer in the two calls.
+The example execution below tricks the software into thinking the user is logged in, by
+taking advantage of the UaF to change the boolean:
+
+```
+> auth admin                                   
+[ auth = 0x78322e1000, service = 0x0 ]       
+> reset                                        
+[ auth = 0x78322e1000, service = 0x0 ]       
+> service aaaaaaaaa0aaaaaaaaa0aaaaaaaaa0121    
+[ auth = 0x78322e1000, service = 0x78322e1000 ]
+> login                                        
+You are already logged in!
+```
+
+In this example, four commands are issued:
+1) 'auth admin'  will allocate the auth\_t structure for the first time. At
+this point, the user has a name, but is not authorized (logged\_in is false).
+2) 'reset'  will free the memory (but pointer to auth is not set to null).
+service (...) will allocate a new string, potentially at the same address
+where the auth\_t structure was previously allocated. 
+3) 'service aaaaaaaaa0aaaaaaaaa0aaaaaaaaa0121' will end up setting the memory
+previously pointing to an auth\_t.
+4) 'login' will use the dangling auth\_t pointer. If this memory has been
+reallocated to the attacker controlled string, it will appear as if the field
+"name" is set to aaaaaaaaa0aaaaaaaaa0aaaaaaaaa012 and the boolean
+logged\_in to true.
+
+:::
+
+Detecting UaFs is usually not an easy task, as it depends not only on
+user inputs, but sometimes also on the execution flow. It can get even more
+complicated in multi-threaded environments. Some UaF detectors intercept calls
```

Comment:
It feels to me like there is a sentence missing here.
The text jumps from "Detecting UaFs is hard" to "some detectors".
I think it would be useful to add a sentence in between that says something like "Therefore, many different UaF detection tools have been built, based on a number of different approaches:"
And then maybe make it a bullet point list or enumerated list, each describing one approach?